File: Problems 40-49/problem41.jl
#Problem 41
using Primes
using BenchmarkTools
"""
____________mas grande podria ser de 9 digitos y tendria que usar del 1 al 9. cuantos primos de 9 digitos hay? no debe haber muchos.
podemos empezar revisando esos.
cuales numeros enteros son  de 9 digitos? los numeros entre 10^8, 10^9-1
"""

number_line = Primes.primes(10^3, 10^8-1)

function removezeroes(x)
    if 0 in digits(x)
        return false
    else
        return true
    end
end

println("Filtering zeroes")
@time nline = Iterators.filter(x -> removezeroes(x), number_line)



function CheckPandigital(x)
    d = Set(digits(x))
    s = string(x)
    l1 = length(d)
    l2 = length(s)
    if l1 == l2
        check = []
        a = 1:l1
        for i in a
            append!(check, i in d)
        end
        if all(check)
            return true
        else
            return false
        end
    else
        return false
    end
end


j = Iterators.filter(x -> CheckPandigital(x), nline)

# @btime maximum(j)
# print(maximum(j))
nums = []
@time begin
    Threads.@threads for i in collect(nline)
        if CheckPandigital(i)
            push!(nums, i)
        end
    end
    println(maximum(nums))
end

File: Problems 40-49/problem41.py
import math
import sympy

nline = sympy.primerange(10**3, 10**8 - 1)

def removezeroes(x):
    if '0' in str(x):
        return False
    return True

nline = filter(removezeroes, nline)

def pandigital(x):
    d = str(x)
    s = set(d)
    l1 = len(d)
    l2 = len(s)
    if l1 == l2:
        c = [lambda y: y in s for y in range(1, l2+1)]
        if all(c):
            return True
        else:
            return False
    else:
        return False

print(max(filter(pandigital, nline)))



